### PR TITLE
Update Rollup and related plugins to their most recent versions

### PR DESCRIPTION
**Pull request by 0xdevalias**
## Summary

Update Rollup and related plugins to their most recent versions + resolve any breaking changes/deprecations/etc along the way. I made each change piece by piece, so the commit history tells a pretty good story of what was changed where/how/why.

fixes https://github.com/facebook/react/issues/24894

For the full deepdive/context, see:

- https://github.com/facebook/react/issues/24894

The inspiration for this came from @jasonwilliams 's PR for attempting to add sourcemap output support to React's builds:

- https://github.com/facebook/react/issues/20186
  - https://github.com/facebook/react/pull/21946

But I figured that it would be useful to minimise the scope of changes in that PR, and to modernise the build tooling along the way.

If any of these updates rely on a node version later than `10.x`, then the following PR may have to land first, otherwise things might break on AppVeyor:

- https://github.com/facebook/react/issues/24891
  - https://github.com/facebook/react/pull/24892

## How did you test this change?

See some of my testing notes RE: figuring compatible package versions + their associated changes in https://github.com/facebook/react/issues/24894#issuecomment-1182841669 Of particular note is the difference in the `react.development.js` output between Rollup `1.x` and rollup `2.x`:

> I'm not currently sure if these changes are significant/change the semantics of the output, or are just a different way of the bundler working, but I did notice that it increases the output file sizes a bit (most dramatically in `react.development.js`):

I ran the tests with `yarn test --prod`, and opened a separate issue for the few tests that seem to fail on Node `16.x`:

- https://github.com/facebook/react/issues/24915

```
⇒  yarn test --prod

..snip..

Test Suites: 3 failed, 2 skipped, 283 passed, 286 of 288 total
Tests:       3 failed, 89 skipped, 7553 passed, 7645 total
Snapshots:   125 passed, 125 total
Time:        245.388 s
Ran all test suites.
```

I'm not 100% certain if this is the most thorough/complete way to test things, so I am definitely open to suggestions on how to improve confidence in these changes.

### CODE REVIEW

File: scripts/rollup/build.js [hunk 1]
```diff
@@ -540,7 +544,7 @@ async function createBundle(bundle, bundleType) {
   const rollupConfig = {
     input: resolvedEntry,
     treeshake: {
-      pureExternalModules,
+      moduleSideEffects: pureExternalModules,
```
Comment (0xdevalias) [inline scripts/rollup/build.js:547]: - https://github.com/rollup/rollup/blob/master/CHANGELOG.md#200
  - > The `treeshake.pureExternalModules` option will now show a deprecation warning when used: use `treeshake.moduleSideEffects: 'no-external'` instead
- https://rollupjs.org/guide/en/#treeshakepureexternalmodules

File: packages/react-devtools-shared/src/hooks/__tests__/updateMockSourceMaps.js [hunk 2]
```diff
@@ -323,7 +323,11 @@ async function bundle() {
     plugins: [
       rollupResolve(),
       commonjs(),
-      babel({presets: ['@babel/preset-react'], sourceMap: true}),
+      babel({
+        babelHelpers: 'bundled',
```
Comment (0xdevalias) [inline packages/react-devtools-shared/src/hooks/__tests__/updateMockSourceMaps.js:327]: - https://github.com/rollup/plugins/blob/master/packages/babel/CHANGELOG.md#500
  - > Removed `externalHelpers` & `runtimeHelpers` options. There is now a single `babelHelpers` option which can take one of `'bundled'`, `'inline'`, `'runtime'` and `'external'` as a value. The default is `'bundled'` which matches `4.x` behavior, **but it is recommended to configure this option explicitly.**
- https://github.com/rollup/plugins/tree/master/packages/babel#babelhelpers

**Review by 0xdevalias — COMMENTED** [review]
Added a few comments/pointers to help explain some of the changes made + point to relevent docs/changelog/etc.

**Comment by 0xdevalias** [comment]
The build is currently failing due to the following error:

```
 BUILDING  react-server-dom-webpack-writer.browser.production.min.server.js (umd_prod)

Could not resolve import "react-server-dom-webpack/src/ReactFlightServerWebpackBundlerConfig" in /home/circleci/project/packages/react-server/src/forks/ReactFlightServerConfig.dom-browser.js using exports defined in /home/circleci/project/packages/react-server-dom-webpack/package.json.

internal/fs/utils.js:314
    throw err;
    ^

Error: ENOENT: no such file or directory, stat 'build/node_modules/react/package.json'
    at Object.statSync (fs.js:1127:3)
    at updatePackageVersions (/home/circleci/project/scripts/rollup/build-all-release-channels.js:261:22)
    at processStable (/home/circleci/project/scripts/rollup/build-all-release-channels.js:127:5)
    at Object.<anonymous> (/home/circleci/project/scripts/rollup/build-all-release-channels.js:66:5)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47 {
  errno: -2,
  syscall: 'stat',
  code: 'ENOENT',
  path: 'build/node_modules/react/package.json'
}
```

I wonder if there could be some `node_modules` cache weirdness happening maybe?

**Comment by poteto** [comment]
@sebmarkbage just noticed these files got deleted, is that an accident or am I missing something?

Screenshot 2023-02-23 at 10 43 06 AM

**Comment by eps1lon** [comment]
@poteto that should've been fixed in https://github.com/facebook/react/pull/26206#issuecomment-1437486474

**Pull request merged**